### ISSUE TITLE
Support parallel decommission with tablets

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylladb#21452]
Tablets enable "parallel bootstrap" where many nodes can be quickly added as cluster members and then receive ownership of data concurrently.

The reverse process is currently not possible, since "decommission" as a topology operation waits for tablet replicas to be moved away from the node, and we serialize topology operations. 

This could be fixed:
  1) Introduce a new per-node flag called `draining`, orthogonal to existing node states, independent of node's state `being_decommissioned`. This will be picked up by tablet scheduler to initiate migrations away from the node and to not allocate new tablets there. Multiple nodes can be marked as `draining` in topology at a time. We cannot reuse `being_decommission` state, because vnode-based token metadata is capable of handling only one node being decommissioned at a time, so decommission in vnode-based keyspaces still needs to be serialized. The `draining` state affects only tablet scheduler.
  1) When decommission is requested, the node is marked as `draining` in topology at the time request is queued, similar to how removenode marks a node as `excluded`. 
  2) tablet scheduler is already capable of draining multiple nodes marked as `being_decommissioned` concurrently, we need to change it to recognize the new `draining` flag. Tablet allocator should also recognize it and not allocate tablets on such nodes.
  3) The draining stage of decommission topology operation will drive tablet movement on behalf of all the draining nodes. We want to keep the "draining" stage so that successful request execution means that tablets were moved away and the node is safe to be physically recycled.
  4) We may change draining stage to exit as soon all tablets of the currently decommissioned node are moved away, as an optimization. 
  4) When decommission topology operation is rolled-back (on error), we set the `draining` flag to false.

We will provide a way for admin to cancel the decommission request if it's not in progress yet. e.g. `nodetool decommission --cancel`, which clears the request and `draining` flag.

  
@bhalevy @avikivity @kbr-scylla

**Comment by kbr-scylla** [comment]
@gleb-cloudius

**Comment by kbr-scylla** [comment]
> We will provide a way for admin to cancel the decommission request if it's not in progress yet. e.g. nodetool decommission --cancel, which clears the request and draining flag.

See also https://github.com/scylladb/scylladb/issues/17514

**Comment by tgrabiec** [comment]
> `being_decommissioned` is a locator::topology thing. What do you propose to do on the topology coordinator layer? We may have additional flag (orthogonal to the state at all) that instruct tablet manager to keep tablets away from those nodes and start remove procedure automatically when draining completes.

I propose to introduce a new per-node flag called `draining`, which will be recognized by tablet scheduler, but not vnode-based topology.

**Comment by swasik** [comment]
@mykaul I put P3 priority here as scale in performance seems much less important than scale out. Please update if you do not agree.

**Comment by bhalevy** [comment]
> [@mykaul](https://github.com/mykaul) I put P3 priority here as scale in performance seems much less important than scale out. Please update if you do not agree.

Why? First, Sluggish scale-in reflects in the bottom line, since one would pay for underutilized resources.
Second, even when scaling out in smaller increments, sometimes we would need to add larger instances and decommission smaller instances we previously added.  If you add 1 node in each rack, you'd need to decommission also 1 node in each rack to compensate, and you want to decommission those nodes in parallel, not serially, as part of incremental scale-out.

**Comment by swasik** [comment]
> Scaling out fast to react quickly to either compute or storage pressure is more important than scale in, I agree therefore it's not as important. If you wish to change instances, you can always scale out with newer stronger instances and then scale-in (remove) those lower spec'ed instances.

And clarifying,  I do not say it is not needed, rather I am putting lower priority here so that we can first focus on scale out and then scale in. Moreover - in most of the cases you will scale in by just one instance in case of utilization decrease. Of course, in case of instances merge we will have to remove 2 instances at once but consolidation logic is optional for Elastic Cloud 1.0 so the priority can be lower.

**Comment by swasik** [comment]
> [@swasik](https://github.com/swasik) I think that P3 is too low, IMO. P2 would be more appropriate for now.

Sure, my main goal was to add some priority whatever we agree on - thanks for the discussion. Replaced.

**Comment by tgrabiec** [comment]
> > Scaling out fast to react quickly to either compute or storage pressure is more important than scale in, I agree therefore it's not as important. If you wish to change instances, you can always scale out with newer stronger instances and then scale-in (remove) those lower spec'ed instances.
> 
> And clarifying, I do not say it is not needed, rather I am putting lower priority here so that we can first focus on scale out and then scale in. 

We already have fast scale-out.

**Comment by swasik** [comment]
> We already have fast scale-out.

I know. But we also have some performance challenges with scale out (e.g. https://github.com/scylladb/scylladb/issues/20921 or https://github.com/scylladb/scylla-enterprise/issues/4504). I do not want to impose anything here, just suggest that from the elastic cloud point of view we should first focus on solving scale out issues and then if we have capacity left scale in.

**Comment by cezarmoise** [comment]
![Image](https://github.com/user-attachments/assets/d19c2d71-5e86-4453-a6fb-5a8fe592d713)

This scenario represents scaling a cluster where each rack has 3 nodes (i4i.large, i4i.xlarge, i4i.2xlarge) and replaces them with a i4i.4xlarge.

It has to do 9 sequential decommissions, taking 2+ hours total.

**Comment by avikivity** [comment]
This problem also affects scale out (the biggest scale-out problem IMO):

1. Storage utilization drops, so the cluster starts scaling in by decommissioning a node
2. Storage utilization increases again, but we can't cancel the decommission and cannot bootstrap a new node in parallel
3. Storage runs out

**Comment by tgrabiec** [comment]
We could allow canceling decommission, it is already cancelled on failure. It's still sub-optimal, because we have to wait for started migrations to finish before bootstrap can begin, and some progress of decommission will be lost.

**Comment by avikivity** [comment]
Sometimes canceling a decommissioning is what we want (if we're decommissioning a small node due to scale in, but the storage utilization started to increase). In this case, losing the progress is actually a good thing, you cancel is better than decommission + bootstrap in parallel.

**Comment by cezarmoise** [comment]
Since RF=racks, would allowing 1 decommission per rack to run in parallel with other racks be easier/faster to implement?

**Comment by bhalevy** [comment]
> Since RF=racks, would allowing 1 decommission per rack to run in parallel with other racks be easier/faster to implement?

No, the main problem is not at this area, but rather in the upper topology state machine layer

**Comment by tzach** [comment]
Should we update the nodetool decommission / API?
https://docs.scylladb.com/manual/stable/operating-scylla/nodetool-commands/decommission.html

We need to document the new option to decommission more than one node at the time
@annastuchlik FYI

**Comment by tgrabiec** [comment]
> Should we update the nodetool decommission / API? https://docs.scylladb.com/manual/stable/operating-scylla/nodetool-commands/decommission.html

It's not required. To decommission in parallel, you invoke separate "nodetool decommission" for each node. And they progress and complete independently.

> 
> We need to document the new option to decommission more than one node at the time [@annastuchlik](https://github.com/annastuchlik) FYI
